packages/react-components/react-tree/src/hooks/useFlatTreeItems.ts

```diff
@@ -0,0 +1,155 @@
+): readonly [FlatTreeProps, LazyFlatTreeItems] {
+  const [openItems, setOpenItems] = useControllableState({
+    state: React.useMemo(() => options?.openItems && ImmutableSet.from(options.openItems), [options?.openItems]),
+    defaultState: React.useMemo(() => options?.defaultOpenItems && ImmutableSet.from(options?.defaultOpenItems), [
+      options?.defaultOpenItems,
+    ]),
+    initialState: ImmutableSet.emptySet,
+  });
+  const onOpenChange = useEventCallback((ev: TreeOpenChangeEvent, data: TreeOpenChangeData) =>
+    setOpenItems(curr => updateOpenItems(data, curr)),
+  );
+
+  const references = React.useMemo(() => createTreeItemPropsRefs(items), [items]);
+
+  const lazyFlatTreeItems = React.useMemo(
+    () =>
+      createLazyFlatTreeItems(references, {
+        filter: item => isTreeItemVisible(item, { openItems, references }),
+      }),
+    [references, openItems],
+  );
+
+  return [{ openItems, onOpenChange }, lazyFlatTreeItems];
+}
+
+function isTreeItemVisible(
+  itemRef: TreeItemPropsReference,
+  {
+    openItems,
+    references,
+  }: {
+    openItems: ImmutableSet<TreeItemId>;
+    references: TreeItemPropsReferences;
+  },
+) {
+  if (itemRef.props['aria-level'] === 1) {
+    return true;
+  }
+  while (itemRef.parentId !== undefined) {
+    if (!openItems.has(itemRef.parentId)) {
+      return false;
+    }
+    itemRef = references.getParent(itemRef);
+  }
+  return true;
+}
+
+function createTreeItemPropsRefs(flatTreeItemProps: FlatTreeItemProps[]): TreeItemPropsReferences {
+  const root: TreeItemPropsReference = {
+    props: { id: 'root', leaf: false, 'aria-level': 0, 'aria-posinset': 0 },
+    childrenSize: 0,
+  };
+  const treeItemPropsRefsPerId = new Map<string, TreeItemPropsReference>();
+  const refs: TreeItemPropsReference[] = [];
+
+  for (let index = 0; index < flatTreeItemProps.length; index++) {
+    const { parentId, ...treeItemProps } = flatTreeItemProps[index];
+
+    const nextPartialItem = flatTreeItemProps[index + 1] as FlatTreeItemProps | undefined;
+    const currentParentRef = parentId ? treeItemPropsRefsPerId.get(parentId) ?? root : root;
+    const isLeaf = nextPartialItem?.parentId !== treeItemProps.id;
+    const currentLevel = (currentParentRef.props['aria-level'] ?? 0) + 1;
+    currentParentRef.childrenSize++;
+
+    const treeItemPropsRef: TreeItemPropsReference = {
+      props: {
+        ...treeItemProps,
+        'aria-level': currentLevel,
+        'aria-posinset': currentParentRef.childrenSize,
+        leaf: isLeaf,
+      },
+      parentId,
+      childrenSize: 0,
+    };
+    treeItemPropsRefsPerId.set(treeItemPropsRef.props.id, treeItemPropsRef);
+    refs.push(treeItemPropsRef);
+  }
+
+  function getParent(itemRef: TreeItemPropsReference): TreeItemPropsReference {
+    const parentRef = itemRef.parentId ? treeItemPropsRefsPerId.get(itemRef.parentId) : root;
+    // This indicates that an item doesn't have a proper parent
+    if (parentRef === undefined) {
```

Comment:
```suggestion
    if (parentRef === undefined && proces.env.NODE_ENV !== 'production') {
```

prod check?

Reply:
Actually this `throw` will be removed on another PR https://github.com/microsoft/fluentui/pull/26978/files#diff-3c1f29aab9555e01ea80d29bb56a231203b0610050ea51442ed9f4eb97914f08R136